icu4c/source/common/wintz.cpp

```diff
@@ -47,77 +48,60 @@ U_NAMESPACE_BEGIN
 U_INTERNAL const char* U_EXPORT2
 uprv_detectWindowsTimeZone()
 {
-    UErrorCode status = U_ZERO_ERROR;
-    char* icuid = nullptr;
-    char dynamicTZKeyName[MAX_TIMEZONE_ID_LENGTH];
-    char tmpid[MAX_TIMEZONE_ID_LENGTH];
-    int32_t len;
-    int id = GEOID_NOT_AVAILABLE;
-    int errorCode;
-    wchar_t ISOcodeW[3] = {}; /* 2 letter ISO code in UTF-16 */
-    char ISOcode[3] = {}; /* 2 letter ISO code in UTF-8 */
-
+    /* Obtain TIME_ZONE_INFORMATION from the API and get the non-localized time zone name. */
     DYNAMIC_TIME_ZONE_INFORMATION dynamicTZI;
     uprv_memset(&dynamicTZI, 0, sizeof(dynamicTZI));
-    uprv_memset(dynamicTZKeyName, 0, sizeof(dynamicTZKeyName));
-    uprv_memset(tmpid, 0, sizeof(tmpid));
 
-    /* Obtain TIME_ZONE_INFORMATION from the API and get the non-localized time zone name. */
-    if (TIME_ZONE_ID_INVALID == GetDynamicTimeZoneInformation(&dynamicTZI)) {
+    DWORD tzIDStatus = GetDynamicTimeZoneInformation(&dynamicTZI);
+    if (tzIDStatus == TIME_ZONE_ID_INVALID || dynamicTZI.TimeZoneKeyName[0] == 0)
         return nullptr;
-    }
 
-    id = GetUserGeoID(GEOCLASS_NATION);
-    errorCode = GetGeoInfoW(id, GEO_ISO2, ISOcodeW, 3, 0);
+    // Is this the correct way to detect cases where the user turns off DST
+    // in Control Panel for a timezone with DST?
+    if (tzIDStatus == TIME_ZONE_ID_UNKNOWN && dynamicTZI.DaylightBias != 0) {
+        LONG utcOffsetMins = dynamicTZI.Bias;
+        if (utcOffsetMins == 0)
+          return uprv_strdup("UTC");
+
+        if (utcOffsetMins % 60 == 0) {
+          char gmtOffsetTz[11]; // "Etc/GMT+dd" is 11-char long with a terminal null.
+          snprintf(gmtOffsetTz, 11, "Etc/GMT%+d", -utcOffsetMins / 60);
+          return uprv_strdup(gmtOffsetTz);
+        }
+    }
 
-    // convert from wchar_t* (UTF-16 on Windows) to char* (UTF-8).
-    u_strToUTF8(ISOcode, UPRV_LENGTHOF(ISOcode), nullptr,
-        reinterpret_cast<const UChar*>(ISOcodeW), UPRV_LENGTHOF(ISOcodeW), &status);
+    // tzIDStatus == TIME_ZONE_ID_STANDARD || tzIDStatus == TIME_ZONE_ID_DAYLIGHT ||
+    //  dynamicTZI.DaylightBias == 0)
 
-    LocalUResourceBundlePointer bundle(ures_openDirect(nullptr, "windowsZones", &status));
-    ures_getByKey(bundle.getAlias(), "mapTimezones", bundle.getAlias(), &status);
+    CharString winTZ;
+    UErrorCode status = U_ZERO_ERROR;
+    winTZ.appendInvariantChars(UnicodeString(TRUE, dynamicTZI.TimeZoneKeyName, -1), status)
+        .append('\0', status);
```

Comment:
checking |status| wouldn't help because appendInvariantChars() just returns |*this| without indicating that EnsureCapacity() failed (i.e. OOM happens). I'll check if winTZ is empty.

Reply:
Hmm, maybe I'm missing something...?
It looks like `ensureCapacity` will set the `errorCode` to `U_MEMORY_ALLOCATION_ERROR` if OOM happens here:
https://github.com/unicode-org/icu/blob/master/icu4c/source/common/charstr.cpp#L198
So shouldn't the status be an error after calling `appendInvariantChars` (if OOM happened)?

Reply:
oops.  I guess that's why I didn't check |status| in my original PR. In that case, I don't have to check |status|. |status| is just percolated down through a chain of ICU APIs,  all of which just return right away without doing anything for U_FAILURE(status). And eventually, the following will take care of any error. 

```
if (U_FAILURE(status)) {
  return nullptr;
}}
```

Reply:
BTW, thank you for catching my blunder with a redundant null termination.